# Recap: You've protected your backpack and made some friends in the forest! Now, one of them comes up to you and gives you a very important dataset. Because as you know, the biggest threat to kitties is not an empty food bowl, but instead...

# DOGS!

First, let's get behind our computer

![](images/20221108102822.png)  

This data might tell us something we can use to defeat the dogs, awesome! Now let's load it.

We are currently in the root of our working directory, the main folder of our adventures.

We have here a data folder with one file in it, "cats_dogs.csv"

First, we'll have to load our pandas army using "import pandas as pd":

Why "pd" you ask? Well simple; if you just "import pandas", you'll have to type pandas full out every time, thats really annoying miauw! so now we can just write pd, here I'll show you:

In [3]:
pd.DataFrame({"Kittens" : [1, 2], "Names" : ["Tessy", "Jonas"]})

NameError: name 'pd' is not defined

You see, I made a dataframe! Awesome. As you can see, it's defined identically as a dictionary.

But instead of a dictionary item, we use pandas columns; "column_name" : [list, of, items]

Let me show you that you can transform a dictionary into a dataframe for our pandas:

In [ ]:
test_dict = {}
test_dict["column one"] = [1, 2]
test_dict["column two"] = [2, 3]

dictionary_dataframe = pd.DataFrame(test_dict)
dictionary_dataframe

The columns before column one is our index list. When we import our dataset, we'll set index=False, so that we don't have this extra column. We'll do this soon.

First, your own dataframe. You can do this either directly like we did at the beginning, or make a dictionary first, whatever you want. 

Give it two columns with two items each. You could make a column with some cat names, and a column with their favourite toys, for instance.

Great, now they can play with their favourite toy!

![](images/20221108103339.png)  

Do you think you are ready for importing the dataset? Let's try it!

I'll write down the PATH to our dataset for you, to make it easier

In [5]:
PATH = "data/cats_dogs.csv"

Now it's up to you to load it. The function is called pd.read_csv()..

Don't forget to add index=False in the function, so we don't get that annoying extra column..

In [6]:
import pandas as pd
cat_dataset = pd.read_csv(PATH)

In [7]:
cat_dataset[36:39]

Number of Households (in 1000) Percentage of households with pets  \
36                           5.172                               56,9   
37                         434.000                               53,0   
38                           1.759                               54,1   

    Number of Pet Households (in 1000) Percentage of Dog Owners  \
36                               2.942                     32,9   
37                             230.000                     29,3   
38                             951.000                     38,6   

    Dog Owning Households (1000s) Mean Number of Dogs per household  \
36                          1.702                               1,5   
37                        127.000                               1,3   
38                        678.000                               1,8   

    Dog Population (in 1000) Percentage of Cat Owners  Cat Owning Households  \
36                     2.485                     33,8                  1.748   
37                   161.000                     27,6                120.000   
38                     1.191                     27,8                490.000   

   Mean Number of Cats  Cat Population  
36                 2,0           3.544  
37                 1,8         212.000  
38                 2,1           1.039

Now let's see out dataset!

In [20]:
cat_dataset

Number of Households (in 1000) Percentage of households with pets  \
0                            1.828                               59,5   
1                            2.515                               59,5   
2                         1148.000                               62,4   
3                           12.974                               52,9   
4                            1.986                               61,3   
5                            1.337                               54,4   
6                          334.000                               56,6   
7                          287.000                               21,9   
8                            7.609                               54,4   
9                            3.798                               55,1   
10                         568.000                               62,0   
11                           5.026                               51,8   
12                           2.478                               59,9   
13                           1.219                               53,6   
14                           1.133                               61,0   
15                           1.777                               61,6   
16                           1.702                               55,1   
17                         548.000                               62,9   
18                           2.169                               52,3   
19                           2.618                               50,4   
20                           3.804                               55,4   
21                           2.163                               53,0   
22                        1115.000                               56,4   
23                           2.498                               61,4   
24                         410.000                               61,3   
25                         710.000                               51,3   
26                         986.000                               55,6   
27                         508.000                               56,8   
28                           3.177                               50,7   
29                         773.000                               67,6   
30                           7.512                               50,6   
31                           3.701                               56,4   
32                         272.000                               53,9   
33                           4.661                               57,4   
34                           1.479                               58,9   
35                           1.505                               63,6   
36                           5.172                               56,9   
37                         434.000                               53,0   
38                           1.759                               54,1   
39                         333.000                               65,6   
40                           2.583                               59,6   
41                           9.002                               58,5   
42                         930.000                               51,2   
43                         265.000                               70,8   
44                           3.017                               53,4   
45                           2.632                               62,7   
46                         765.000                               62,1   
47                           2.350                               57,5   
48                         221.000                               61,8   

    Number of Pet Households (in 1000) Percentage of Dog Owners  \
0                                1.088                     44,1   
1                                1.497                     40,1   
2                              716.000                     47,9   
3                                6.865                     32,8   
4                 

Wooooow that's wayy to much information, let's use a cool function called head(), which shows the number of ROWS that we want to see, by default this is 5.

.head() does not require "pd.", as .head() works directly on a dataset.. Google if you have to.

Also, since there is a "." before "head()", you know that this is a method, and that means it is APPLIED to a variable. So it comes after the variable, and not before. .head(cat_dataset) is WRONG.

Beautiful! Now we can see it clearly. You know what, I don't like to scroll, let's just print out all the columns of this pandas dataset directly, can you do that for me?

Wow.. that's a lot to work with. Why don't we take the first column "Number of Households (in 1000)" and print only that column.

Remember, pandas dataframes work very much like dictionaries, and can be accessed the same way. Try to print out the first column "Number of Households (in 1000)" here:

PRO TIP: instead of writing down the entire name of the column, maybe you could also just put in the first index of the list of columns we made above? Would be easier, right? something like .columns[0]?

In [29]:
num_households = ...
print(num_households)

pandas.core.series.Series

Pfew, so we have 48 rows in our dataset. Interesting. So the kitties here tell me that these are all different districts in our kitty wonderland, and they all have their own row. So the 0'th district has 1.828 * 1000 households with kitties in them, so cute!

![](images/20221108104427.png)  

Why don't we try and get the average of all the districts? I think we should ask some help from our friends over at Numpy, they are really good with numbers!

import numpy as np    below!

In [30]:
import numpy as np

Great, now let's use our np.mean() function to print our mean number of households

Just 218?? I thought more. Anyway, let's go back to our entire cat_dataframe.. Why don't you print the rows 36-39. Pandas dataframes work like lists in that sense

You see that! District 38 has a massive number of households, let's take that row!

In [ ]:
distric_38 = 

print(distric_38)

But why didnt we just use cat_dataset[38]? Well, indexing works out somewhat differently from lists for instance. With lists, you can give the list[index] and you'll get that item, but pandas dataframes work like dictionaries: dataframe[item], 

where item stands for column name

and our district 38 is a row at index 38, we know that.

so the our_dataframe.iloc[index] method gives us the ROW at that index

Just for fun, let's try and compare the number of households against the SUM of all households in our dataset. The sum function is done identically to the numpy's mean function, try it out.

To be clear, you want to devide district 38's number of households to the TOTAL sum of households of the dataset, so you'll have to devide here and print it

In [ ]:
print(..)

Great!

Suddenly many well-dressed kitties barge in through the door..

![](images/20221108105837.png)  

Since you've learned the .iloc method and the dataframe indexing, we DEMAND you make a new dataset 

containing only the rows between 5 and 40!

In [ ]:
demanded_dataset = ...
demanded_dataset.head(10)

Okay okay, I did.. what now angry kitties!?

"Now PRINT using a for loop, all the percentage of dog owners for every district! We need to see the truth!"

We'll help you a bit. we'll print all the district indexes for you, all 34 (between 5 and 40) of the demanded dataset. Use the district index to print the percentage of dog owners for each district!

hinthint: .iloc[...][column name]

In [ ]:
for district in range(len(demanded_dataset)):
    print(district)



Some kitties seemed to have left, the pain was too deep...

![](images/20221108110730.png)  

"Okay.. Okay, just like we feared. Fine!" The lead kitty looks distressed..

"Take our previous for loop that we'll add below, and this time print for every district if the percentage of dog owners is greater than the percentage of cat owners."

You can do this with an if-else statement, either printing TRUE or FALSE.. or you can be smart and use boolean operation:

```python
a = 10
b = 20

if a > b:
    print(True)
else:
    print(False)
```

Or.. you can just use boolean operation directly:

```python
print(a > b)
```

In [ ]:
for district in range(len(demanded_dataset)):
    ...

    
    # Hint: uncomment the following code if you're in trouble
    # current_district = demanded_dataset.iloc[district]
    # percentage_dog_owners = ...
    # percentage_cat_owners = ...
    # print(percentage_dog_owners > percentage_cat_owners)

Oh MIAUWSELS! It seems many districts have more dig owners than cat owners... Thats terrible! Let's use our dataframe knowledge next time to find out why, and how to stop it!

![](images/20221108111724.png)  